concepts/decorators/about.md

```diff
@@ -1,2 +1,246 @@
+They usually return either the wrapped function or the result of an expression that uses the wrapped function.
+
+A simple decorator - one that simply returns its wrapped function - can be written as follows:
+```python
+>>> def do_nothing(func):
+...     return func
+...
+... @do_nothing
+... def function4():
+...     return 4
+...
+>>> print(function4())
+4
+```
+
+A decorator may only add side effects, such as additional information used for logging:
+
+```python
+>>> def my_logger(func):
+...     print(f"Entering {func.__name__}")
+...     return func
+...
+... @my_logger
+... def my_func():
+...     print("Hello")
+...
+>>> my_func()
+Entering my_func
+Hello
+```
+
+A decorator does not return itself.
+It may return its function arguments, another function, or one or more values that replace the return from the passed-in or decorated function.
+If a decorator returns another function, it will usually return an [inner function][inner functions].
+
+## Inner Functions
+
+A function can be defined within a function.
+Such a nested function is called an [inner function][inner functions].
+A decorator may use an inner function to wrap its function argument.
+The decorator then returns its inner function.
+The inner function may then return the original function argument.
+
+### A Validating Decorator Using an Inner Function
+
+Following is an example of a decorator being used for validation:
+
+```python
+>>> def my_validator(func):
+...     def my_wrapper(world):
+...         print(f"Entering {func.__name__} with {world} argument")
+...         if ("Pluto" == world):
+...             print("Pluto is not a planet!")
+...         else:
+...             return func(world)
+...     return my_wrapper
+...
+... @my_validator
+... def my_func(planet):
+...     print(f"Hello, {planet}!")
+...
+>>> my_func("World")
+Entering my_func with World argument
+Hello, World!
+...
+>>> my_func("Pluto")
+Entering my_func with Pluto argument
+Pluto is not a planet!
+```
+
+On the first line, we have the definition for the decorator with its `func` argument.
+On the next line is the definition for the decorators _inner function_, which wraps the `func` argument.
+Since the _inner function_ wraps the decorator's `func` argument, it is passed the same argument that is passed to `func`.
+Note that the wrapper doesn't have to use the same name for the argument that was defined in `func`.
+The original function uses `planet` and the decorator uses `world`, and the decorater still works.
+
+The inner function returns either `func` or, if `planet` equals `Pluto`, it will print that Pluto is not a planet.
+It could be coded to raise a `ValueError` instead.
+So, the inner function wraps `func`, and returns either `func` or does something that substitutes what `func` would do.
+The decorator returns its _inner function_.
+It does not _directly_ return the original, passed-in function.
+Depending on what happens in the wrapper function, `func` may or may not be returned.
```

Comment:
This does not, for me, clarify the prior sentence.

Reply:
The whole paragraph will be rephrased in the next commit to

The inner function returns either `func` or, if `planet` equals `Pluto`, it will print that Pluto is not a planet.
It could be coded to raise a `ValueError` instead.
So, the inner function wraps `func`, and returns either `func` or does something that substitutes what `func` would do.
The decorator returns its _inner function_.
The _inner_function_ may or may not return the original, passed-in function.
Depending on what code conditionally executes in the wrapper function or _inner_function_, `func` may be returned, an error could be raised, or a value of `func`s return type could be returned.


It states and then repeats itself to hopefully be more reinforcing than obfuscatory.